File: python/robot-name/robot_name.py
from secrets import choice
from string import ascii_uppercase, digits


class Robot(object):
    names_in_use = set()

    def __init__(self):
        self.name = ""
        self.new_name()

    def reset(self):
        old_name = self.name
        self.new_name()
        # Only remove the old name after the new name has been generated to
        # stop the very small chance that the same name is generated again.
        Robot.names_in_use.remove(old_name)

    def new_name(self):
        random_name = Robot.generate_random_name()

        while random_name in Robot.names_in_use:
            random_name = Robot.generate_random_name()

        Robot.names_in_use.add(random_name)
        self.name = random_name

    @staticmethod
    def generate_random_name():
        random_letters = ''.join([choice(ascii_uppercase) for _ in range(2)])
        random_digits = ''.join([choice(digits) for _ in range(3)])
        random_name = random_letters + random_digits
        return random_name

File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;
using System.Text;

public class Robot
{
    private static readonly Random random = new Random();
    private static readonly HashSet<string> NamesInUse = new HashSet<string>();

    public Robot() => Reset();

    public string Name { get; private set; }

    public void Reset()
    {
        string new_name;
        do
        {
            var sb = new StringBuilder();
            sb.Append((char) random.Next('A', 'Z'));
            sb.Append((char) random.Next('A', 'Z'));
            sb.Append(random.Next(0, 9));
            sb.Append(random.Next(0, 9));
            sb.Append(random.Next(0, 9));
            new_name = sb.ToString();
        } while (NamesInUse.Contains(new_name));

        NamesInUse.Remove(Name);
        Name = new_name;
        NamesInUse.Add(Name);
    }
}